packages/trpc/server/routers/viewer/slots/handleNotificationWhenNoSlots.ts

```diff
@@ -0,0 +1,123 @@
+import { Redis } from "@upstash/redis";
+
+import type { Dayjs } from "@calcom/dayjs";
+import { sendOrganizationAdminNoSlotsNotification } from "@calcom/emails";
+import { IS_PRODUCTION, WEBAPP_URL } from "@calcom/lib/constants";
+import { getTranslation } from "@calcom/lib/server";
+import { prisma } from "@calcom/prisma";
+
+type EventDetails = {
+  username: string;
+  eventSlug: string;
+  startTime: Dayjs;
+  visitorTimezone?: string;
+  visitorUid?: string;
+};
+
+// 7 days or 60s in dev
+const NO_SLOTS_NOTIFICATION_FREQUENCY = IS_PRODUCTION ? 604_800 : 60;
+
+const NO_SLOTS_COUNT_FOR_NOTIFICATION = 2;
+
+const constructRedisKey = (eventDetails: EventDetails, orgSlug?: string) => {
+  return `${eventDetails.username}:${eventDetails.eventSlug}${orgSlug ? `@${orgSlug}` : ""}`;
+};
+
+const constructDataHash = (eventDetails: EventDetails) => {
+  const obj = {
+    startTime: eventDetails.startTime.format("YYYY-MM-DD"),
+    visitorTimezone: eventDetails?.visitorTimezone,
+    visitorUid: eventDetails?.visitorUid,
+  };
+
+  return JSON.stringify(obj);
+};
+
+export const handleNotificationWhenNoSlots = async ({
+  eventDetails,
+  orgDetails,
+}: {
+  eventDetails: EventDetails;
+  orgDetails: { currentOrgDomain: string | null; isValidOrgDomain: boolean };
+}) => {
+  // Check for org
+  if (!orgDetails.currentOrgDomain) return;
+  const UPSTASH_ENV_FOUND = process.env.UPSTASH_REDIS_REST_URL && process.env.UPSTASH_REDIS_REST_TOKEN;
+  if (!UPSTASH_ENV_FOUND) return;
+
+  // Check org has this setting enabled
+  const orgSettings = await prisma.team.findFirst({
+    where: {
+      slug: orgDetails.currentOrgDomain,
+      isOrganization: true,
+    },
+    select: {
+      organizationSettings: {
+        select: {
+          adminGetsNoSlotsNotification: true,
+        },
+      },
+    },
+  });
+
+  if (!orgSettings?.organizationSettings?.adminGetsNoSlotsNotification) return;
+
+  const redis = Redis.fromEnv();
+
+  const usersUniqueKey = constructRedisKey(eventDetails, orgDetails.currentOrgDomain);
+  // Get only the required amount of data so the request is as small as possible
+  // We may need to get more data and check the startDate occurrence of this
+  // Not trigger email if the start months are the same
+  const usersExistingNoSlots = await redis.lrange(usersUniqueKey, 0, NO_SLOTS_COUNT_FOR_NOTIFICATION - 1);
+  await redis.lpush(usersUniqueKey, constructDataHash(eventDetails));
+
+  if (!usersExistingNoSlots.length) {
+    await redis.expire(usersUniqueKey, NO_SLOTS_NOTIFICATION_FREQUENCY);
+  }
```

Comment:
If this is the first push to the key - we push a expiry date (to the key not each item in the array)